### ISSUE TITLE
`Style/RedundantCondition` is doing unsafe corrections

### ISSUE THREAD

**Issue by byroot** [rubocop/rubocop#13913]
After https://github.com/rubocop/rubocop/pull/13729 the `Style/RedundantCondition`  is doing corrections that I don't believe are safe.

One example from the Rails codebase:

```ruby
     def save(**) # :nodoc:
-      Suppressor.registry[self.class.name] ? true : super
+      Suppressor.registry[self.class.name] || super
     end
```

This change the code from returning `true` to return anything that is "thruthy", which in some case could be a problem.

**Comment by bbatsov** [comment]
Yeah, this needs stricter checks (e.g. that condition is a predicate) as noted by the author of the PR. @datpmt Would you be able to look into fixing the issue? 

I'm still concerned that given Ruby's track record with predicates (not all of them return booleans) this might be hard to enforce reliably, so we might consider a configuration option or reverting the problematic change.

**Comment by datpmt** [comment]
@bbatsov

Let me take a look at the issue. If we can't address it in time for the upcoming release, we can revert the change. What do you think?

**Comment by nevans** [comment]
Related to the issue as reported:

* `foo bar: errors[name.to_sym] && "error"`
  is replaced with `foo bar: errors[name.to_sym]`
* `"fmt string %s%s" % [foo, baz && " (#{bar})"]`
  is replaced with `"fmt string %s%s" % [foo, baz]`

In both cases, the value of the expression is used for more than just a boolean check.  This is fine for conditionals (`if bool`, `elsif bool`, `unless bool`, `while bool`, `until bool`, `bool ? expr1 : expr2`), but for anything else we may care about the distinction between `nil` vs `false` or `true` vs other truthy values.

**Comment by nevans** [comment]
I was assuming that these were related, but I wasn't paying close attention and my issue is coming from `Lint/LiteralAsCondition`, not `Style/RedundantCondition`.  I'll create a separate issue for mine.

**Comment by koic** [comment]
>  If we can't address it in time for the upcoming release, we can revert the change. What do you think?

I've opened #13919 to resolve this issue.

**Comment by a-lavis** [comment]
It doesn't seem like this is fixed in 1.73.1 for me:

Given the following `.rubocop.yml`:
```yml
AllCops:
  NewCops: enable
  TargetRubyVersion: 3.4
```
and the following `foo.rb`:
```ruby
# frozen_string_literal: true

def foo? = 123

puts({ foo: foo? ? true : false })
```

If you run `ruby foo.rb`, you get `{foo: true}`.

Then if you run `rubocop -a foo.rb`, it changes it to:
```ruby
# frozen_string_literal: true

def foo? = 123

puts({ foo: foo? || false })
```

Then, if you run `ruby foo.rb`, you instead get `{foo: 123}`.

So it seems like this is still not a safe autocorrect.

`rubocop -V`:
```
1.73.1 (using Parser 3.3.7.1, rubocop-ast 1.38.1, analyzing as Ruby 3.4, running on ruby 3.4.2) [arm64-darwin24]
```
